stackit/internal/services/iaas/securitygrouprule/resource_test.go

```diff
@@ -0,0 +1,294 @@
+				Direction:             types.StringNull(),
+				Description:           types.StringNull(),
+				EtherType:             types.StringNull(),
+				IpRange:               types.StringNull(),
+				RemoteSecurityGroupId: types.StringNull(),
+				IcmpParameters:        types.ObjectNull(icmpParametersTypes),
+				PortRange:             types.ObjectNull(portRangeTypes),
+				Protocol:              types.ObjectNull(protocolTypes),
+			},
+			true,
+		},
+		{
+			"simple_values",
+			Model{
+				ProjectId:           types.StringValue("pid"),
+				SecurityGroupId:     types.StringValue("sgid"),
+				SecurityGroupRuleId: types.StringValue("sgrid"),
+			},
+			&iaasalpha.SecurityGroupRule{
+				Id:                    utils.Ptr("sgrid"),
+				Description:           utils.Ptr("desc"),
+				Direction:             utils.Ptr("ingress"),
+				Ethertype:             utils.Ptr("ether"),
+				IpRange:               utils.Ptr("iprange"),
+				RemoteSecurityGroupId: utils.Ptr("remote"),
+				IcmpParameters:        &fixtureIcmpParameters,
+				PortRange:             &fixturePortRange,
+				Protocol:              &fixtureProtocol,
+			},
+			Model{
+				Id:                    types.StringValue("pid,sgid,sgrid"),
+				ProjectId:             types.StringValue("pid"),
+				SecurityGroupId:       types.StringValue("sgid"),
+				SecurityGroupRuleId:   types.StringValue("sgrid"),
+				Direction:             types.StringValue("ingress"),
+				Description:           types.StringValue("desc"),
+				EtherType:             types.StringValue("ether"),
+				IpRange:               types.StringValue("iprange"),
+				RemoteSecurityGroupId: types.StringValue("remote"),
+				IcmpParameters:        fixtureModelIcmpParameters,
+				PortRange:             fixtureModelPortRange,
+				Protocol:              fixtureModelProtocol,
+			},
+			true,
+		},
+		{
+			"empty_port_range",
+			Model{
+				ProjectId:           types.StringValue("pid"),
+				SecurityGroupId:     types.StringValue("sgid"),
+				SecurityGroupRuleId: types.StringValue("sgrid"),
+			},
+			&iaasalpha.SecurityGroupRule{
+				Id:        utils.Ptr("sgrid"),
+				PortRange: &iaasalpha.PortRange{},
+			},
+			Model{
+				Id:                    types.StringValue("pid,sgid,sgrid"),
+				ProjectId:             types.StringValue("pid"),
+				SecurityGroupId:       types.StringValue("sgid"),
+				SecurityGroupRuleId:   types.StringValue("sgrid"),
+				Direction:             types.StringNull(),
+				Description:           types.StringNull(),
+				EtherType:             types.StringNull(),
+				IpRange:               types.StringNull(),
+				RemoteSecurityGroupId: types.StringNull(),
+				IcmpParameters:        types.ObjectNull(icmpParametersTypes),
+				PortRange: types.ObjectValueMust(portRangeTypes, map[string]attr.Value{
+					"max": types.Int64Null(),
+					"min": types.Int64Null(),
+				}),
+				Protocol: types.ObjectNull(protocolTypes),
+			},
+			true,
+		},
+		{
+			"empty_protocol",
```

Comment:
what this test case should cover is if the user has set an empty object but the API returns nil, which should not throw value conversion error, so it should be the following: 
```
Model{
				ProjectId:           types.StringValue("pid"),
				SecurityGroupId:     types.StringValue("sgid"),
				SecurityGroupRuleId: types.StringValue("sgrid"),
                                Protocol: types.ObjectValueMust(protocolTypes, map[string]attr.Value{
					"name":   types.StringNull(),
					"number": types.Int64Null(),
				})
			},
			&iaasalpha.SecurityGroupRule{
				Id:       utils.Ptr("sgrid"),
			},
			Model{
				Id:                    types.StringValue("pid,sgid,sgrid"),
				ProjectId:             types.StringValue("pid"),
				SecurityGroupId:       types.StringValue("sgid"),
				SecurityGroupRuleId:   types.StringValue("sgrid"),
				Direction:             types.StringNull(),
				Description:           types.StringNull(),
				EtherType:             types.StringNull(),
				IpRange:               types.StringNull(),
				RemoteSecurityGroupId: types.StringNull(),
				IcmpParameters:        types.ObjectNull(icmpParametersTypes),
				PortRange:             types.ObjectNull(portRangeTypes),
				Protocol: types.ObjectValueMust(protocolTypes, map[string]attr.Value{
					"name":   types.StringNull(),
					"number": types.Int64Null(),
				}),
			},
```

Reply:
If protocol needs to have name or number, this case will be obsolete, right

Reply:
yes, not relevant anymore

Reply:
removed test